src/server/requestHandlers.js

```diff
@@ -1,431 +1,490 @@
+        var story = storyDoc.value;
+
+        var contentTypes = {};
+
+        fluid.each(story.content, function (contentBlock) {
+            contentTypes[contentBlock.blockType] = true;
+        });
+
+        story.contentTypes = fluid.keys(contentTypes);
+
+        story = fluid.censorKeys(story, ["content"]);
+
+        storyBrowse.stories[storyDoc.id] = story;
+
+    });
+
+    return storyBrowse;
+};
+
+
+fluid.defaults("sjrk.storyTelling.server.getStoryHandler", {
+    gradeNames: "kettle.request.http",
+    invokers: {
+        handleRequest: {
+            funcName: "sjrk.storyTelling.server.handleGetStory",
+            args: ["{request}", "{server}.storyDataSource", "{server}.options.secureConfig.uploadedFilesHandlerPath"]
+        }
+    }
+});
+
+sjrk.storyTelling.server.handleGetStory = function (request, dataSource, uploadedFilesHandlerPath) {
+    var id = request.req.params.id;
+    var promise = dataSource.get({directStoryId: id});
+
+    promise.then(function (response) {
+
+        fluid.transform(response.content, function (block) {
+            if (block.blockType === "image") {
+                if (block.imageUrl) {
+                    block.imageUrl = uploadedFilesHandlerPath + "/" + block.imageUrl;
+                }
+                return block;
+            } else if (block.blockType === "audio" || block.blockType === "video") {
+                if (block.mediaUrl) {
+                    block.mediaUrl = uploadedFilesHandlerPath + "/" + block.mediaUrl;
+                }
+                return block;
+            }
+        });
+
+        var responseAsJSON = JSON.stringify(response);
+        request.events.onSuccess.fire(responseAsJSON);
+    }, function (error) {
+        var errorAsJSON = JSON.stringify(error);
+        request.events.onError.fire({
+            isError: true,
+            message: errorAsJSON
+        });
+    });
+};
+
+fluid.defaults("sjrk.storyTelling.server.saveStoryWithBinariesHandler", {
+    gradeNames: "kettle.request.http",
+    requestMiddleware: {
+        "saveStoryWithBinaries": {
+            middleware: "{server}.saveStoryWithBinaries"
+        }
+    },
+    invokers: {
+        handleRequest: {
+            funcName: "sjrk.storyTelling.server.handleSaveStoryWithBinaries",
+            args: ["{arguments}.0", "{server}.storyDataSource", "{server}.options.globalConfig.authoringEnabled"]
+        }
+    }
+});
+
+sjrk.storyTelling.server.handleSaveStoryWithBinaries = function (request, dataSource, authoringEnabled) {
+    if (authoringEnabled) {
+        var rotateImagePromises = [];
+
+        // rotate any images based on their EXIF data, if present
+        fluid.each(request.req.files.file, function (singleFile) {
+            if (singleFile.mimetype && singleFile.mimetype.indexOf("image") === 0) {
+                var singleRotatePromise = sjrk.storyTelling.server.rotateImageFromExif(singleFile).then(null, function (error) {

```

Comment:
If you attach `then()` here, the promise returned by `rotateImageFromExif` will be resolved here. It doesn't make sense to push `singleRotatePromise` into an array and use `Promise.all` or `fluid.promise.sequence` later to wait for their executions because at then promises in the array have already been resolved here. If you do expect to have one promise returned when all `singleRotatePromise` are executed, remove the `then()` resolver from here and attach resolve/reject functions on the promise returned by `fluid.promise.sequence`.